components/tools/OmeroJava/test/integration/UpdateServiceTest.java

```diff
@@ -1997,4 +2006,71 @@ public void testCreateTagSetAnnotationUsingPojo() throws Exception {
         Assert.assertNotNull(data.getNameSpace());
         Assert.assertEquals(data.getNameSpace(), TagAnnotationData.INSIGHT_TAGSET_NS);
     }
+
+    /**
+     * Test that users cannot adjust the {@link ExternalInfo} of their own session.
+     * @throws Exception unexpected
+     */
+    @Test
+    public void testEditSessionExternalInfo() throws Exception {
+        newUserAndGroup("rw----");
+        /* prepare to act as root */
+        final IQueryPrx iQueryRoot = root.getSession().getQueryService();
+        final IUpdatePrx iUpdateRoot = root.getSession().getUpdateService();
+        /* as root, add external info with a specific UUID to the user's current session */
+        Session session = (Session) iQueryRoot.findByString("Session", "uuid", client.getSessionId());
+        final long sessionId = session.getId().getValue();
+        final ExternalInfo infoOld = new ExternalInfoI();
+        infoOld.setEntityType(omero.rtypes.rstring(Session.class.getName()));
+        infoOld.setEntityId(omero.rtypes.rlong(sessionId));
+        infoOld.setUuid(omero.rtypes.rstring(UUID.randomUUID().toString()));
+        session.getDetails().setExternalInfo(infoOld);
+        iUpdateRoot.saveObject(session);
+        /* prepare for querying session with external info */
+        final String hql = "FROM Session s JOIN FETCH s.details.externalInfo WHERE s.id = :id";
+        final Parameters params = new ParametersI().addId(sessionId);
+        /* as the normal user, check that the session external info UUID is as set by root */
+        session = (Session) iQuery.findByQuery(hql, params);
+        ExternalInfo info = session.getDetails().getExternalInfo();
+        Assert.assertEquals(info.getUuid().getValue(), infoOld.getUuid().getValue());
+        /* as the normal user, attempt to set the session external info to something else */
+        final ExternalInfo infoNew = new ExternalInfoI();
+        infoNew.setEntityType(omero.rtypes.rstring(Session.class.getName()));
+        infoNew.setEntityId(omero.rtypes.rlong(sessionId));
+        infoNew.setUuid(omero.rtypes.rstring(UUID.randomUUID().toString()));
+        Assert.assertNotEquals(infoOld.getUuid().getValue(), infoNew.getUuid().getValue());
+        session.getDetails().setExternalInfo(infoNew);
+        try {
+            iUpdate.saveObject(session);
+            Assert.fail("the user cannot change the external info of a session");
+        } catch (ValidationException ve) {
+            /* expected */
+        }
+        /* as the normal user, check that the session external info UUID remains as set by root */
+        session = (Session) iQuery.findByQuery(hql, params);
+        info = session.getDetails().getExternalInfo();
+        Assert.assertEquals(info.getUuid().getValue(), infoOld.getUuid().getValue());
+        /* as the normal user, attempt to set the session external info UUID to something else */
+        info.setUuid(infoNew.getUuid());
+        iUpdate.saveObject(info);
```

Comment:
So `saveObject` on the modified `ExternalInfo` object is expected to fail 'silently' (ie no exception thrown), in contrast to saving a session with new `ExternalInfo` object? Just wonder what the reason for this difference is.